packages/taiko-client/proposer/profitability.go

```diff
@@ -0,0 +1,389 @@
+func (p *Proposer) getPercentageFromBaseFeeToTheProposer(num *big.Int) *big.Int {
+	if p.protocolConfigs.BaseFeeConfig().SharingPctg == 0 {
+		return big.NewInt(0)
+	}
+
+	result := new(big.Int).Mul(num, big.NewInt(int64(p.protocolConfigs.BaseFeeConfig().SharingPctg)))
+	return new(big.Int).Div(result, big.NewInt(100))
+}
+
+// txWithBaseFee is a helper struct to sort transactions by their base fee
+type txWithBaseFee struct {
+	tx      *types.Transaction
+	baseFee *big.Int
+	gas     uint64
+}
+
+// findOptimalBaseFeeThreshold finds the optimal base fee threshold that maximizes collected fees.
+// This algorithm has complexity:
+// - O(n) to flatten and extract base fees
+// - O(n log n) to sort transactions by base fee
+// - O(n^2) to find the optimal threshold
+// Total: O(n^2) where n is the number of transactions
+// Returns the optimal base fee and the collected fees at that threshold.
+func (p *Proposer) findOptimalBaseFeeThreshold(txBatch []types.Transactions) (*big.Int, *big.Int) {
+	// Flatten all transactions into a single slice with their base fees
+	var allTxs []txWithBaseFee
+	for _, txs := range txBatch {
+		for _, tx := range txs {
+			txBaseFee := tx.GasFeeCap()
+			if txBaseFee == nil {
+				txBaseFee = tx.GasPrice()
+			}
+			if txBaseFee != nil {
+				allTxs = append(allTxs, txWithBaseFee{
+					tx:      tx,
+					baseFee: new(big.Int).Set(txBaseFee),
+					gas:     tx.Gas(),
+				})
+			}
+		}
+	}
+
+	if len(allTxs) == 0 {
+		return nil, big.NewInt(0)
+	}
+
+	// Sort transactions by base fee in descending order (highest first)
+	sort.Slice(allTxs, func(i, j int) bool {
+		return allTxs[i].baseFee.Cmp(allTxs[j].baseFee) > 0
+	})
+
+	baseFeeSharePct := p.protocolConfigs.BaseFeeConfig().SharingPctg
+	var bestBaseFee *big.Int
+	bestCollectedFees := big.NewInt(0)
+	cumulativeGas := uint64(0)
+
+	// Now iterate through sorted transactions to find optimal threshold
+	// Key insight: if we set base fee to tx[i].baseFee, we include tx[0..i]
+	// Collected fees = (baseFee * proposerSharePct + avgTip) * sum(gas[0..i])
+	for i := 0; i < len(allTxs); i++ {
+		cumulativeGas += allTxs[i].gas
+		// Calculate collected fees if we use this base fee as threshold
+		// All transactions from 0 to i have baseFee >= currentBaseFee
+		currentBaseFee := allTxs[i].baseFee
+
+		// Calculate total revenue from base fees (proposer share)
+		baseFeeForProposer := new(big.Int).Mul(
+			currentBaseFee,
+			big.NewInt(int64(baseFeeSharePct)),
+		)
+		baseFeeForProposer.Div(baseFeeForProposer, big.NewInt(100))
+		baseFeeRevenue := new(big.Int).Mul(
+			baseFeeForProposer,
+			new(big.Int).SetUint64(cumulativeGas),
+		)
+
+		// Calculate total revenue from tips for all included transactions
+		tipRevenue := big.NewInt(0)
+		for j := 0; j <= i; j++ {
+			txTip := p.calculatePriorityFee(allTxs[j].tx, currentBaseFee)
+			txTipRevenue := new(big.Int).Mul(txTip, new(big.Int).SetUint64(allTxs[j].gas))
+			tipRevenue.Add(tipRevenue, txTipRevenue)
+		}
```

Comment:
If optimizing this is really critical, we can discuss some approaches. But it seems fine right now.